### DISCUSSION TITLE
Leaving a trail on a PIXI.Graphics

### DISCUSSION THREAD

**Discussion by BlueNebulaDev** [pixijs/pixijs#7814] [Q&A] [answered]
Every frame of my simulation I'm drawing a line on a PIXI.Graphics without clearing it. I'd like to see the evolution of the line visually. After a few hundred frames, PIXI.Graphics starts slowing down. After thousands it becomes very slow.

My theory is that it re-renders all the lines every time I modify it. If this is the case, would it be possible to remove the older lines I drew and only leave a buffer of say the latest 100?

**Comment by ivanpopelyshev** [comment, marked as answer]
> My theory is that it re-renders all the lines every time I modify it.

existance of function `clear()` should help with that. As for "leave latest 100" , well, you can just trim `graphics.geometry.graphicsData` and call `graphics.geometry.invalidate()` every time you do something strange.

I understand that for you PixiJS is black box, without sources. Here's my advice: its very hard to do WebGL-related stuff without looking in sources. There's no book on performance, there's not even collection articles, and even more - there's no single article on performance. All you can find is answer on issues and html5gamedevs forum posts. Its even worse - its not just pixi problem - its for all webgl renderers. You either look in sources and learn stuff, or do things randomly and hope it works

**Reply by BlueNebulaDev** [reply]
Thank you, with `geometry.graphicsData` I managed to only keep the last 100.

You're right about looking at the sources... PIXI is quite a large piece of software though and I find it quite hard to figure out where to look at.

Few days ago I tried to look at the sources because of another issues I ran into: my viewport was zoomed in quite a lot and a length of 1 took around 500px; circles with a small radius (for instance 0.5 radius) drawn in PIXI.Graphics and SmoothGraphics were very boxy: they were approximated by a dozen of lines or so. I spent half an hour looking in the code for the function that draws circles, but couldn't find it, so at the end I decided to scale up the coordinates I feed to PIXI.

**Reply by ivanpopelyshev** [reply]
Clone PixiJS (or another webgl renderer) locally. Open in separate IDE window (vscode or idea/webstorm), press specific shortcut to search for class or filename. If you use something (Sprite, Graphics, SimpleMesh, SimpleRope) and aren't sure about specific functions/fields, just open corresponding file. That's much faster than googling.

That way you'll see which parts are magic and which are very stupid logic that you can alter from outside depending on your need.
